sphinx_immaterial/mermaid_diagrams.py

```diff
@@ -0,0 +1,59 @@
+"""A custom directive that allows using mermaid diagrams"""
+from typing import List
+from docutils import nodes
+from docutils.parsers.rst import directives
+from sphinx.util.docutils import SphinxDirective
+from sphinx.application import Sphinx
+
+
+class mermaid_node(nodes.General, nodes.Element):
+    pass
+
+
+class MermaidDirective(SphinxDirective):
+    """a special directive"""
+
+    has_content = True
+    option_spec = {
+        "name": directives.unchanged,
+        "class": directives.class_option,
+    }
+
+    def run(self) -> List[nodes.Node]:
+        """Run the directive."""
+        self.assert_has_content()
+        content = "\n".join(self.content)
+        diagram = mermaid_node("", classes=["mermaid"], content=content)
+        diagram += nodes.literal("", content, format="html")
+        diagram_div = nodes.container(
+            "",
+            is_div=True,
+            classes=["mermaid-diagram"] + self.options.get("class", []),
+        )
+        if self.options.get("name", ""):
+            self.add_name(diagram_div)
+        diagram_div += diagram
+        self.set_source_info(diagram_div)
+        return [diagram_div]
+
+
+def visit_mermaid_node_html(self, node: mermaid_node):
+    attributes = {"class": "mermaid"}
+    self.body.append(self.starttag(node, "pre", **attributes))
+
+
+def depart_mermaid_node(self, node: mermaid_node):
+    self.body.append("</pre>")
+
+
+def visit_mermaid_node_latex(self, node: mermaid_node):
+    self.body.append('<pre class="mermaid">')
```

Comment:
This doesn't work, right?

This theme isn't intended to be used except with the html builder, so I don't think we need to worry about latex.

Reply:
It does work with latex writer. That's why I have the CI building both html and latex. In fact the RTD builds are making PDFs from the latex builds.

The mermaid feature won't work with latex since it relies on JS. In fact if the JS in a browser isn't loaded, then the mermaid directives look like non-highlighted code blocks. This is what I'm aiming for with the latex writer.

This function is important because the `starttag()` used in html doesn't exist in the latex builder.

Reply:
Oh I see ---  putting the mermaid code literally in the output seems reasonable enough for latex (rendering the diagrams would be better but of course that would be a lot more trouble to accomplish), but why include HTML like `<pre class="mermaid">`?

Reply:
that's just what it does for html. I don't really know latex syntax...

Actually the mermaid directive results in a
```html
<div class="mermaid-diagram">
  <pre class="mermaid"> <!--this part is reuired for the mermaid.js to work-->
    <code>
      <!-- all special markup language characters are converted -->
    </code>
  </pre>
</div>
```

Reply:
I installed all the `tex-*` stuff used in the RTD docker and test built a pdf from the latex output... The HTML syntax used here (for latex writers) is now replaced with latex syntax to better imitate a literal code block in the generated PDF.

As a side note, I don't think that PDFs are built on RTD because of other numerous unknown references (seems related to `.. _ref_id:` tactic). A PDF can be built from this theme's latex output, but you have to accept that there's no ToC written. We can address this in the future if someone is so inclined, but it isn't the primary goal of this theme. So, for now, I'm just doing the bare minimum to allow a latex builder to complete its job.